components/shared/CustomButton.tsx

```diff
@@ -6,7 +6,7 @@ interface CustomButtonProps extends ButtonProps {
   children?: React.ReactNode;
   text?: string;
 }
-const CustomButton = ({ className, text }: CustomButtonProps) => {
+const CustomButton = ({ className, text, children }: CustomButtonProps) => {
   return <Button className={twMerge(`${className}, 'gap-4'`)}>{text}</Button>;
 };
```

Comment:
We've added children but we aren't using them. This feels like it might be a mistake. If you want to render the children you need to use them inside the return:

```suggestion
const CustomButton = ({ className, text, children }: CustomButtonProps) => {
  return <Button className={twMerge(`${className}, 'gap-4'`)}>{text} {children}</Button>;
};
```
You might not want 'text' now though, if the way we're rendering our buttons is to use the children.

We'd probably want this, and then we'd want to rework any instances of `CustomButton`
```js
const CustomButton = ({ className, children }: CustomButtonProps) => {
  return <Button className={twMerge(`${className}, 'gap-4'`)}>{children}</Button>;
};
```

I'm a little confused about the point of the CustomButton if all it does is behave the same as "Button", but not pass in any props `{...props}` for onClick etc... We probably want to be able to use the button's functions:

```js
interface CustomButtonProps{
    children: React.ReactNode;
    className: string;
    props: any;
}
 const CustomButton = ({ children,className,props }: CustomButtonProps) =>{
    return <Button {...props} className={`${className} gap-4`}>{children}</Button>;
  };

```

It looks like this is just `Button` from shadcn with a `p-4`, if we're planning on making this with a bunch of custom styles that's great! But then why are we not completing it and putting it in it's own "reusable components" PR?

Reply:
im actually confused as to why i need this button at all, isn't the shadcn-ui button by default a custom button, the Button in the components folder does the exact aame stuff. I had made it when doing the auth cause you had mentioned making a reusable button when i was doing the initial auth stuff that i didnt know I wasnt supposed to start.  It's not adding anything that the button is doing, and it isn't working when Im doing what I'm trying to do with the side bar.  I was rushing to catch up with everything cause it seemed like i was falling behind where i was supposed to be even tho I was ahead.  So at worst it seems like an issue that can be scrapped for now and revisted in the future when i do catch up, especially if It's not adding anything different to the component that I can tell.

Reply:
```
 the Button in the components folder does the exact aame stuff
 ```
 
 Yep! That's the issue I have with it. The idea behind creating our custom reusable components is to make our code cleaner and easier to work with. 
 
 I was not very clear with the suggestion, and I assumed you understood what I meant. That's my mistake. Let's clear it up.
 
Let's forget about this incomplete component in this PR. Let's instead have a `reusable components` PR after that will give us a bunch of easy-to-use components for our application.
 
Take a look at the design and take note of all the different button designs. We have:

1) A Login Button
2) Some Social Login Buttons with Icons
3) A Next Button
4) An 'add goal checkbox' Button with an Icon
5) A `create post` gradient button with an Icon
6) A `create post` button without a gradient
7) A `new resource button` with an icon
8) An `Add Checkmark` button with an icon
9) A 'ghost' variant Logout button
10) An Add New Link button with Icon
 ...etc...
 
 Our goal should be to create a reusable component that will let us quickly create these buttons inside of our app without littering our JSX with a ton of repeated code.
 
 Instead of continuing to write (and this is pseudo-code written in this editor) stuff like this, including imports and repeated styling and repeated images.:
 ```jsx
 import {plusIcon} from "myicons";
 import Image from "next/image";
import {Button} from "components/ui";

 ...etc...
 
 <Button variant={'secondary'} className="flex justify-center items-center gap-x-2 shadow-sm w-full text-etc...">
 <Image src={plusIcon} alt="plus-icon" width={10} height={10} />
 Add checkmark
 </Button>
 // ...etc...
  <Button variant={'secondary'} className="flex gap-x-2  justify-center items-center shadow-sm w-full text-etc...">
 <Image src={plusIcon} alt="plus-icon" width={10} height={10} />
 New Resource
 </Button>
 
 <Button variable={'primary'} className="flex justify-center items-center shadow-sm w-full text-etc...">
 Create Post
 </Button>
 
 ```
 
 It would be great if we could have a clean reusable component that can take a single line.
 
 This benefits us because we can focus on the content of the page as we develop further- instead of matching the design each time making sure we're consistently using the same text sizes, colors, flex gaps, image paddings, image sizings, hover effects, etc... on every single button we write.
 
 This is great for when we want to make improvements or adjustments as well. Let's say halfway through our development or maybe even after- we decide we want a completely different Icon for the "plus" button. Or maybe we want it to look drastically different. 
 
 If you're just using the Button from shadcn and manually adding the Image to every component or page that needs the button to be used: you need to go back into every file that uses that button and make adjustments as needed. If you miss one: now you have inconsistent design.
 
 We should plan ahead- and make reusable components.
 
 You could start with a base Button (again this is an example only):
 ```js

 import { Button } from "@/components/ui/button";

// take the prop type from a regular button, now our button expects onClicks, children, etc... 
interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
// anything else you need to add can go in here
}

 const OurButton = ({ children, ...props }: ButtonProps) => {
  return <Button {...props}>{children}</Button>;
};

export default OurButton;
```

Then maybe we could add our 'plus icon' variant here:
```js
OurButton.Plus = function OurButtonPlus({ children, ...props }: ButtonProps) {
  return (
    <OurButton {...props}>
      <PlusIcon />
      {children}
    </OurButton>
  );
};
```

And we could structure our styles to easily editable and readable from the top, and add them into our components classes:

```js
const baseClass = "px-4 py-2 rounded-md text-white bg-slate-500";
const variants = {
  plus: `hover:bg-red-500 gap-x-2`,
  social: `bg-secondary-500 hover:bg-secondary-600 text-lg text-black-500`,
  primary: `bg-danger-500 hover:bg-danger-600 text-xl text-white-500`,
  gradient: `bg-gradient-to-r from-primary-500 to-primary-600 text-xl text-white-500`,
};

const OurButton = ({ children, ...props }: ButtonProps) => {
  const combinedClass = `${baseClass} ${props.className}`;
  return (
    <Button {...props} className={combinedClass}>
      {children}
    </Button>
  );
};

OurButton.Plus = function OurButtonPlus({ children, ...props }: ButtonProps) {
  return (
    <OurButton {...props} className={variants["plus"]}>
      <PlusIcon />
      {children}
    </OurButton>
  );
};

export default OurButton;

```

You can use it simply with:
```js
import Button from "../shared/Button";
...etc...
    <Button.Plus>Plus Button</Button.Plus>
```

Take a look at the `button.tsx` source from shadcn. You could instead do a lot of the styles with variants, but you might want to extend the functionality anyway. Maybe we want the option to have icons on the right, or buttons with other elements built in etc...


The same idea can be done with our other reusable components. 
 
 We have these tags:
 
![image](https://github.com/agordon123/git_note/assets/4108484/2ae6a432-08ad-497e-902e-a9896d7ae864)


It would be nice to simply have a component we could import for consistent styling instead of writing this component/button over and over again in every page.

Reply:
Ok that's a good explanation, yeah I misunderstood the original request and I remember during a session we had talked specifically about using the custom button so when I tried to use it for the buttons in the sidebar it wasnt working, so my mind was in 10 places at once when I was doing it.  The way I'm going to go about finishing the rest of the layout (the routing part we talked about and that's how I have it in the subbranch I had made but didn't publish yet).  

The only other note I have is I want to hear from your end, what are you expecting of me when I open a pull request on code that is incomplete?  should I make the feature branch and then open up sub-branches for every component/component group?  That seems like a solution, but yeah I think we should clear this up ASAP before i keep doing it.